platform/pulpcore/app/models/storage.py

```diff
@@ -1,400 +1,110 @@
-        """
-        super(FileContent, self).__init__(io.StringIO(content), '-')
-        self.content = content
-        self.size = len(content)
-
-    def open(self, mode='r'):
-        """
-        Open for reading.
-
-        Args:
-            mode (str): See: open() mode.
-        """
-        if 'b' in mode:
-            self.file = io.BytesIO(bytearray(self.content, encoding='utf8'))
-        else:
-            self.file = io.StringIO(self.content)
-
-    def __bool__(self):
-        return len(self.content) > 0
-
-
-@deconstructible
-class TLSLocation:
-    """
-    Determine storage location (path) for PEM encoded TLS keys and certificates
-    associated with any model.
-
-    Attributes:
-        name (str): The file name.
-    """
-
-    MEDIA_TYPE = 'tls'
-
-    def __init__(self, name):
-        """
-        Args:
-            name (str): The file name.
-        """
-        self.name = name
-
-    def __call__(self, model, name):
+    def get_available_name(self, name, max_length=None):
         """
-        Determine storage location as: MEDIA_ROOT/tls/<model>/<id>/<name>.
+        Returns a filename if a file by that name does not exist
 
         Args:
-            model (pulpcore.app.models.Model): The model object.
-            name (str): The (unused) input file path.
+            name (string): Requested file name
+            max_length (int): Maximum length of the filename. Not used in this implementation.
 
         Returns:
-            str: An absolute (base) path
-        """
-        return os.path.join(
-            settings.MEDIA_ROOT,
-            self.MEDIA_TYPE,
-            type(model).__name__,
-            str(model.id),
-            self.name)
-
-
-class FileSystem(Storage):
-    """
-    Provides simplified filesystem storage.
-    The django filesystem storage is overly complex and makes incompatible assumptions
-    and has incompatible behaviors with regard to how final storage paths are calculated.
-    """
-
-    @staticmethod
-    def unlink(path):
-        """
-        Delete the link at the specified path.
-        No exception is raised if the link does not exist.
-
-        Args:
-            path (str): Path to unlink.
-
-        Raises:
-            OSError: On error.
-        """
-        try:
-            os.unlink(path)
-        except FileNotFoundError:
-            pass
-
-    @staticmethod
-    def copy(path, content):
-        """
-        Atomic copy of the content into storage.
-        The content copied into a temporary file, then renamed.
-        This prevents creating partial files.
-
-        Args:
-            path(str): The storage path.
-            content (File): Source file object.
+            Name of the file.
 
         Raises:
-            FileExistsError: File already stored.
-            OSError: On error.
+            OSError if a file with requested name already exists.
         """
-        tmp_path = '.'.join((path, str(uuid4()), 'tmp'))
```

Comment:
This tmp file logic ensures that a file being copied into storage is atomic.  This ensures that if the process is killed, it cannot leave a half-baked file in place.  If half-baked, the logic will think the file already exists in subsequent save requests.  Best I can tell, the FileSystemStorage copy is not atomic.  Am I missing something?